src/operator/contrib/csr_neighborhood_sample-inl.h

```diff
@@ -163,96 +493,266 @@ static void GetSample(std::vector<dgl_id_t>& ver_list,
+  size_t num_seeds = seed_arr.shape().Size();
+  CHECK_GE(max_num_vertices, num_seeds);
+
+  const dgl_id_t* val_list = csr.data().dptr<dgl_id_t>();
+  const dgl_id_t* col_list = csr.aux_data(csr::kIdx).dptr<dgl_id_t>();
+  const dgl_id_t* indptr = csr.aux_data(csr::kIndPtr).dptr<dgl_id_t>();
+  const dgl_id_t* seed = seed_arr.data().dptr<dgl_id_t>();
+  dgl_id_t* out = sampled_ids.data().dptr<dgl_id_t>();
 
   // BFS traverse the graph and sample vertices
   dgl_id_t sub_vertices_count = 0;
-  std::unordered_map<dgl_id_t, bool> sub_ver_mp;
+  std::unordered_set<dgl_id_t> sub_ver_mp;
   std::queue<ver_node> node_queue;
   // add seed vertices
-  for (size_t i = 0; i < seed_num; ++i) {
+  for (size_t i = 0; i < num_seeds; ++i) {
     ver_node node;
     node.vertex_id = seed[i];
     node.level = 0;
     node_queue.push(node);
-    sub_ver_mp[node.vertex_id] = true;
+    sub_ver_mp.insert(node.vertex_id);
     sub_vertices_count++;
   }
 
   std::vector<dgl_id_t> tmp_src_list;
-  std::vector<dgl_id_t> tmp_sampled_list;
+  std::vector<dgl_id_t> tmp_edge_list;
+  std::vector<dgl_id_t> tmp_sampled_src_list;
+  std::vector<dgl_id_t> tmp_sampled_edge_list;
+  std::unordered_map<dgl_id_t, neigh_list> neigh_mp;
 
-  while (!node_queue.empty() && sub_vertices_count < max_num_vertices) {
+  size_t num_edges = 0;
+  while (!node_queue.empty()) {
     ver_node& cur_node = node_queue.front();
     if (cur_node.level < num_hops) {
+
       dgl_id_t dst_id = cur_node.vertex_id;
       tmp_src_list.clear();
-      tmp_sampled_list.clear();
-      GetSrcList(col_list, indptr, dst_id, tmp_src_list);
-      GetSample(tmp_src_list, num_neighbor, tmp_sampled_list);
-      for (size_t i = 0; i < tmp_sampled_list.size(); ++i) {
-        auto got = sub_ver_mp.find(tmp_sampled_list[i]);
-        if (got == sub_ver_mp.end()) {
+      tmp_edge_list.clear();
+      tmp_sampled_src_list.clear();
+      tmp_sampled_edge_list.clear();
+
+      GetSrcList(val_list, 
+                 col_list, 
+                 indptr, 
+                 dst_id, 
+                 tmp_src_list, 
+                 tmp_edge_list);
+
+      if (probability == nullptr) {  // uniform-sample
+        GetUniformSampleReplace(tmp_src_list, 
+                       tmp_edge_list, 
+                       num_neighbor, 
+                       tmp_sampled_src_list,
+                       tmp_sampled_edge_list);
+      } else {  // non-uniform-sample
+        GetNonUniformSample(probability,
+                       tmp_src_list, 
+                       tmp_edge_list, 
+                       num_neighbor, 
+                       tmp_sampled_src_list,
+                       tmp_sampled_edge_list);
+      }
+
+      neigh_mp.insert(std::pair<dgl_id_t, neigh_list>(dst_id,
+        neigh_list(tmp_sampled_src_list, tmp_sampled_edge_list)));
+      num_edges += tmp_sampled_src_list.size();
+      
+      // TODO The code doesn't limit the maximal number of vertices correctly.
+      sub_vertices_count++;
```

Comment:
sub_ver_mp contains the sampled vertices. Do we need count the number of sampled vertices?

it's a bit tricky when limiting the number of sampled vertices. actually, we need to make sure `sub_ver_mp.size() <= max_num_vertices`.